rpc/handlers.go

```diff
@@ -1174,3 +1155,91 @@ func (h *Handler) EstimateMessageFee(msg MsgFromL1, id BlockID) (*FeeEstimate, *
 	}
 	return &estimates[0], nil
 }
+
+// TraceTransaction returns the trace for a given executed transaction, including internal calls
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/1ae810e0137cc5d175ace4554892a4f43052be56/api/starknet_trace_api_openrpc.json#L11
+func (h *Handler) TraceTransaction(hash felt.Felt) (json.RawMessage, *jsonrpc.Error) {
+	_, blockHash, blockNumber, err := h.bcReader.Receipt(&hash)
+	if err != nil {
+		return nil, ErrTxnHashNotFound
+	}
+
+	block, err := h.bcReader.BlockByNumber(blockNumber)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	pendingBlock := blockHash == nil
+
+	state, closer, err := h.bcReader.StateAtBlockHash(block.ParentHash)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer h.callAndLogErr(closer, "Failed to close state in starknet_traceTransaction")
+
+	var (
+		headState       core.StateReader
+		headStateCloser blockchain.StateCloser
+	)
+	if pendingBlock {
+		headState, headStateCloser, err = h.bcReader.PendingState()
+	} else {
+		headState, headStateCloser, err = h.bcReader.HeadState()
+	}
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+	defer h.callAndLogErr(headStateCloser, "Failed to close head state in starknet_traceTransaction")
+
+	var txIndex int
+	var classes []core.Class
+	for i, transaction := range block.Transactions {
+		if transaction.Hash().Equal(&hash) {
+			txIndex = i
+		}
+
+		declareTx, ok := transaction.(*core.DeclareTransaction)
+		if !ok {
+			continue
+		}
+
+		class, stateErr := headState.Class(declareTx.ClassHash)
+		if stateErr != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, stateErr.Error())
+		}
+		classes = append(classes, class.Class)
+	}
+
+	if pendingBlock {
+		height, hErr := h.bcReader.Height()
+		if hErr != nil {
+			return nil, ErrBlockNotFound
+		}
+		blockNumber = height + 1
+	}
+
+	header := block.Header
+
+	sequencerAddress := header.SequencerAddress
+	if sequencerAddress == nil {
+		sequencerAddress = core.NetworkBlockHashMetaInfo(h.network).FallBackSequencerAddress
+	}
+
+	_, traces, err := h.vm.Execute(block.Transactions, classes, blockNumber, header.Timestamp,
+		sequencerAddress, state, h.network, []*felt.Felt{})
+	if err != nil {
+		rpcErr := *ErrContractError
+		rpcErr.Data = err.Error()
+		return nil, &rpcErr
+	}
+	trace := traces[txIndex]
```

Comment:
Do we need to re-execute the whole block, or just `block.Transactions[:txIndex]`?

Reply:
> or just block.Transactions[:txIndex]?

Yeap, nice catch.

Reply:
That would make sense to me. If class isn't present in the state before `txIndex`, we shouldn't need to include it.

Reply:
> we shouldn't need to include it.

added